File: src/main/java/leet_code/WordSearch.java
package leet_code;

import java.util.Arrays;

public class WordSearch {

    public static void main(String[] args) {
//        System.out.println(exist(new char[][]{{'A','B','C','E'},{'S','F','C','S'},{'A','D','E','E'}},"ABCCED"));
        System.out.println(exist(new char[][]{{'A','B','C','E'},{'S','F','C','S'},{'A','D','E','E'}},"ABCB"));
//        System.out.println(exist(new char[][]{{'a'}},"a"));
    }


    public static boolean  exist(char[][] board, String word) {

        int runner = 0;
        boolean[][] visited = new boolean[board.length][board[0].length];

        for (int i = 0; i < board.length; i++) {
            for (int j = 0; j < board[0].length; j++) {
                boolean res = backtracking(board, visited, word, runner, i, j);
                if (res) return true;
            }
        }
        return false;
    }

    private static int[][] directions = new int[][]{{0, 1}, {1, 0}, {-1, 0}, {0, -1}};

    private static boolean backtracking(char[][] board, boolean[][] visited, String word, int runner, int i, int j) {
        if (runner == word.length()) return true;
        if (word.charAt(runner) != board[i][j] || visited[i][j]) {
            return false;
        }

        visited[i][j] = true;
        System.out.println(Arrays.deepToString(visited));
        runner++;
        for (int[] direction : directions) {
            if (moveAble(board, direction, i, j)) {
                if (backtracking(board, visited, word, runner,
                        i + direction[0], j + direction[1])) {
                    return true;
                }
            }
        }
        visited[i][j]=false;
        return false;
    }

    private static boolean moveAble(char[][] board, int[] direction, int i, int j) {
        return i + direction[0] >= 0 && i + direction[0] < board.length &&
                j + direction[1] >= 0 && j + direction[1] < board[0].length;
    }

}

File: src/main/java/leet_code/till_april/WordSearch.go
package main

import (
	"fmt"
)

func main() {

	board := [][]byte{
		{'A', 'B', 'C', 'E'},
		{'S', 'F', 'C', 'S'},
		{'A', 'D', 'E', 'E'},
	}

	fmt.Println(exist(board, "ABCCED"))
}

func exist(board [][]byte, word string) bool {

	i := 0
	j := 0
	runner := 0
	var visited [][]bool
	for i := 0; i < len(board); i++ {
		var arr []bool
		for j := 0; j < len(board[0]); j++ {
			arr = append(arr, false)
		}
		visited = append(visited, arr)
	}

	fmt.Println(visited[0][0])
	return search(board, word, visited, i, j, runner)
}

var directions = [][]int{{0, 1}, {1, 0}, {-1, 0}, {0, -1}}

func search(board [][]byte, word string, visited [][]bool, i, j, runner int) bool {
	fmt.Println(i, j)
	fmt.Println(visited[0][0])
	if runner == len(word) {
		return true
	}
	if board[i][j] == word[runner] {
		runner++
	} else {
		runner = 0
	}

	for _, dir := range directions {
		if moveable(dir, board, i, j) {
			if !visited[i][j] {
				visited[i][j] = true
				res := search(board, word, nil, i+dir[0], j+dir[1], runner)
				if res {
					return res
				}
			}
		}
	}
	visited[i][j] = false
	return false
}

func moveable(dir []int, board [][]byte, i, j int) bool {
	if (i+dir[0] < len(board) && i+dir[0] >= 0) &&
		j+dir[1] < len(board[0]) && j+dir[1] > 0 {
		return true
	}
	return false
}
